hbase-balancer/src/main/java/org/apache/hadoop/hbase/master/balancer/CandidateGenerator.java

```diff
@@ -42,13 +42,13 @@
    */
   int pickRandomRegion(BalancerClusterState cluster, int server, double chanceOfNoSwap) {
     // Check to see if this is just a move.
-    if (cluster.regionsPerServer[server].length == 0
+    if (cluster.regionsPerServer.get(server).size() == 0
         || ThreadLocalRandom.current().nextFloat() < chanceOfNoSwap) {
       // signal a move only.
       return -1;
     }
-    int rand = ThreadLocalRandom.current().nextInt(cluster.regionsPerServer[server].length);
-    return cluster.regionsPerServer[server][rand];
+    int rand = ThreadLocalRandom.current().nextInt(cluster.regionsPerServer.get(server).size());
+    return cluster.regionsPerServer.get(server).get(rand);
```

Comment:
I have to keep RegionsPerServer as ArrayList<ArrayList<Integer>> because we need to get random element while making it compatible with regionsPerHost/Rack.